pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/region_dedicated_numa_exclusive.go

```diff
@@ -86,3 +81,54 @@ func (r *QoSRegionDedicatedNumaExclusive) TryUpdateProvision() {
 		internal.updateStatus = types.PolicyUpdateSucceeded
 	}
 }
+
+func (r *QoSRegionDedicatedNumaExclusive) getControlKnobs() types.ControlKnob {
+	reclaimedCPUSize := 0
+	if reclaimedInfo, ok := r.metaReader.GetPoolInfo(state.PoolNameReclaim); ok {
+		for _, numaID := range r.bindingNumas.ToSliceInt() {
+			reclaimedCPUSize += reclaimedInfo.TopologyAwareAssignments[numaID].Size()
+		}
+	}
+	cpuRequirement := r.ResourceUpperBound + r.ReservedForReclaim - float64(reclaimedCPUSize)
+
+	return types.ControlKnob{
+		types.ControlKnobNonReclaimedCPUSize: {
+			Value:  cpuRequirement,
+			Action: types.ControlKnobActionNone,
+		},
+	}
+}
+
+func (r *QoSRegionDedicatedNumaExclusive) getIndicators() types.Indicator {
+	var (
+		cpi float64
+		err error
+	)
+
+	// get cpi of the main container in this region
+	for podUID, containerSet := range r.podSet {
+		for containerName := range containerSet {
+			ci, ok := r.metaReader.GetContainerInfo(podUID, containerName)
+			if !ok || ci == nil {
+				klog.Errorf("[qosaware-cpu] illegal container info of %v/%v", podUID, containerName)
+				return nil
+			}
+			if ci.ContainerType == v1alpha1.ContainerType_MAIN {
+				cpi, err = r.metaReader.GetContainerMetric(podUID, containerName, consts.MetricCPUCPIContainer)
+				if err != nil {
+					klog.Errorf("[qosaware-cpu] get %v of %v/%v failed: %v", consts.MetricCPUCPIContainer, podUID, containerName, err)
+					return nil
+				}
+			}
+		}
+	}
+
+	return types.Indicator{
+		consts.MetricCPUCPIContainer: {
+			Current: cpi,
```

Comment:
and, should we emit error if there is no valid pod or container (i.e, cpi = 0 or something like that)?

Reply:
Policy will handle all kinds of abnormal cases and write error logs